### PR TITLE
lang/funcs: Add camelCase, kebab-case and snake_case functions

### PR DESCRIPTION
**Pull request by stephen-harris**

### CODE REVIEW

**Comment by pranaygp** [comment]
I would love this @teamterraform :)

**Comment by dhirschfeld** [comment]
This PR has been languishing for a long time now 😞

Could someone from @teamterraform please comment on whether or not this contribution will be accepted if the conflicts are resolved?

File: lang/funcs/string.go [hunk 1]
```diff
@@ -46,8 +46,85 @@ var ReplaceFunc = function.New(&function.Spec{
...
+func SnakeCase(str cty.Value) (cty.Value, error) {
+	return SnakeCaseFunc.Call([]cty.Value{str})
+}
+
+// KebabCase is a Function that converts a given string to kebab-case.
+func KebabCase(str cty.Value) (cty.Value, error) {
+	return KebabCaseFunc.Call([]cty.Value{str})
+}
+
+// KebabCase is a Function that converts a given string to kebab-case.
```
Comment (sikachu) [inline lang/funcs/string.go:127]: Sorry for jumping in, but I just noticed this typo in the comment

```suggestion
// CamelCase is a Function that converts a given string to CamelCase.
```

Comment (solowk) [reply lang/funcs/string.go:127]: // CamelCase is a Function that converts a given string to **c**amelCase.

Comment (sikachu) [reply lang/funcs/string.go:127]: Fixed, thank you!

**Comment by dnck** [comment]
Bumping! Would be nice to have this in terraform :)

**Comment by shayan-golafshani** [comment]
Bumping again! I'm new to TF and this would be great to have.

**Comment by felipefrocha** [comment]
This wasn't released yet right?

**Comment by crw** [comment]
This PR is in a holding pattern for now, please see this comment: https://github.com/hashicorp/terraform/issues/28855#issuecomment-856345294, specifically:

> We're being pretty cautious about adding new functions that overlap with existing use-cases because Terraform has grown quite an unwieldy collection of builtins over the years, and so we've been considering various ways to allow for externally-defined functions (e.g. https://github.com/hashicorp/terraform/issues/2771) to avoid continuing to grow that set. However, that's not something we'll be working on in the very near future, due to our current focus being elsewhere, so we'll revisit this at a later time and post any updates in those other issues.

Just an FYI as to why this hasn't been reviewed. The conversation is active but there are no updates as to externally-defined functions. Once that has been designed, it will be easier to determine which of the existing requests for functions belong in core versus belong outside of core.

**Comment by Shrirang97** [comment]
@teamterraform any update on this?

**Comment by kennedy** [comment]
bump?

**Comment by fardarter** [comment]
Is there are reason this hasn't been reviewed?

**Comment by crw** [comment]
@Shrirang97 no updates at this time.

**Comment by jonmcewen** [comment]
@crw @teamterraform 
 9 months on; what are your latest thoughts on this please?

**Comment by crw** [comment]
@jonmcewen https://github.com/hashicorp/terraform/pull/27357#issuecomment-1093269075

For what it is worth, I am continuing to champion the extendable functions concept. However, there is no update to share.

**Comment by crw** [comment]
Just FYI, https://github.com/hashicorp/terraform/issues/27696#issuecomment-1758977989

**Comment by mcm-ham** [comment]
Is anyone interested in making a provider for these functions? https://github.com/hashicorp/terraform/issues/2771

**Comment by OJFord** [comment]
I haven't tried it or looked at the code yet, so I'm not vouching for completeness or anything, but there is: https://registry.terraform.io/providers/northwood-labs/corefunc/latest/docs/functions/str_snake

**Comment by crw** [comment]
Thanks for sharing that! I am going to close this PR due to age and that it is unlikely to be accepted as a core built-in function. Please feel free to make a case for it, although given that it has not been merged by the maintainers in four years, it would likely need to be a uniquely compelling case.

**Pull request closed without merging**